### PR TITLE
Add initial barcode validation and tests

### PR DESCRIPTION
**Pull request by harikt**
Hi @weierophinney ,

This PR contains Barcode validation .

Some of the tests are commented out which need to be addressed. I am still thinking how we can resolve those.

- [x] Are you creating a new feature?
  - [ ] Why is the new feature needed? What purpose does it serve?
  - [x] Add tests for the new feature.
  - [x] Add documentation for the new feature.
  - [ ] Add a `CHANGELOG.md` entry for the new feature.

### CODE REVIEW

File: src/Barcode.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
+<?php
+declare(strict_types=1);
```
Comment (harikt) [inline src/Barcode.php]: I wonder why `declare(strict_types=1);` is omitted on other classes. Is that purposeful ?

File: test/BarcodeTest.php [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+    {
+        $barcode = new Barcode($adapter);
+        $this->assertInstanceOf($expectedResult, $barcode->getAdapter());
+    }
+
+    // public function testNoneExisting()
+    // {
+    //     $this->expectException(InvalidArgumentException::class);
+    //     $this->expectExceptionMessage('not found');
+    //     $barcode = new Barcode('\Zend\Validate\BarcodeTest\NonExistentClassName');
```
Comment (harikt) [inline test/BarcodeTest.php]: In the old code this send an InvalidArgumentException before calling `getAdapter()`  . Do we want to do the same ? In that case the adapter is created first even if it may / may not is used.

File: test/BarcodeTest.php [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+    public function testArrayConstructAdapter()
+    {
+        $barcode = new Barcode('Ean13', false);
+        $this->assertInstanceOf(Ean13::class, $barcode->getAdapter());
+        $this->assertFalse($barcode->useChecksum());
+    }
+
+
+
+    // public function testDefaultArrayConstructWithMissingAdapter()
```
Comment (harikt) [inline test/BarcodeTest.php]: I think we can remove these tests. What do you think ?

File: test/BarcodeTest.php [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+
+    /**
+     * @group ZF-10116
+     */
+    public function testArrayLengthMessage()
+    {
+        $barcode = new Barcode('ean8');
+        $result = $barcode->validate('123');
+        $this->validateResult($result, false);
+        // $message = $barcode->getMessages();
```
Comment (harikt) [inline test/BarcodeTest.php]: `$result->getMessages()` returns array of `ValidationFailureMessage` . If we can keep the key as `code` of `ValidationFailureMessage` it may be easy to check this. May be there is another way to resolve this.

**Comment by weierophinney** [comment]
@,hatikt - Thanks for starting this! However, per the RFC, this is one of the validators that will get a bridge component, due to the required dependency it had on zend-barcode. If you want to create that component, you can do so under your own user/org, and then open an issue here (or, once we've moved the component to the zendframework org, on that repo) letting us know, so we can bring it into the zendframework org.

The new component would depend on this one and zend-barcode.

File: src/Barcode.php (hunk 1 above)
Comment (weierophinney) [reply src/Barcode.php]: We haven't yet decided if we will require that.

File: test/BarcodeTest.php (hunk 3 above)
Comment (weierophinney) [reply test/BarcodeTest.php]: Yes, as these ports should already the constructor to accept specific arguments, not an option array.

File: test/BarcodeTest.php (hunk 4 above)
Comment (weierophinney) [reply test/BarcodeTest.php]: The code is part of the validation failure messages.

**Comment by harikt** [comment]
Hi @weierophinney , 

Thank you for your quick reply. I am a bit confused for I didn't noticed any dependency on zend-barcode when writing a test for the Barcode validation component.

I will look more into the discussion on https://github.com/zendframework/zend-validator/issues/1 and https://discourse.zendframework.com/t/rfc-new-validation-component/208/2 .

**Comment by weierophinney** [comment]
> I am a bit confused for I didn't noticed any dependency on zend-barcode when writing a test for the Barcode validation component.

Okay, so it turns out I was wrong! In reviewing the various barcode-related validators, none of them have a dependency on zend-barcode. Additionally, it's not a suggested (or required!) dependency either.

So... keep up the great work, and I'll review when you've finished!

**Comment by harikt** [comment]
@weierophinney I have updated the code. I belive you can do a final review.

File: src/Barcode/Royalmail.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        if ($value[0] == '(') {
+            $value = substr($value, 1);
+
+            if ($value[strlen($value) - 1] == ')') {
+                $value = substr($value, 0, -1);
+            } else {
+                return false;
+            }
+        }
```
Comment (weierophinney) [inline src/Barcode/Royalmail.php]: Let's rewrite the above:

```php
if ($value[0] === '(') {
    if ($value[strlen($value) - 1] !== ')') {
        return false;
    }

    $value = substr($value, 1, -1);
}
```

File: src/Barcode/Upce.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+     * @param string $value Value
+     * @return bool
+     */
+    public function hasValidLength($value)
+    {
+        if (strlen($value) != 8) {
+            $this->useChecksum(false);
+        } else {
+            $this->useChecksum(true);
+        }
```
Comment (weierophinney) [inline src/Barcode/Upce.php]: This can be much simpler:

```php
$this->setUseChecksum(strlen($value) === 8);
```

File: test/BarcodeTest.php [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+
+namespace ZendTest\DataValidator;
+
+use PHPUnit\Framework\TestCase;
+use Zend\DataValidator\Barcode;
+use Zend\DataValidator\Result;
+use Zend\DataValidator\Exception\InvalidArgumentException;
+use Zend\DataValidator\Barcode\AdapterInterface;
+use Zend\DataValidator\Barcode\Ean13;
+use Zend\DataValidator\Barcode\Issn;
```
Comment (weierophinney) [inline test/BarcodeTest.php]: Please alphabetize these.

File: test/BarcodeTest.php [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+
+/**
+ * \Zend\Barcode
+ *
+ * @group      Zend_Validator
+ */
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This docblock is irrelevant, and can be removed.

File: test/BarcodeTest.php [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+class BarcodeTest extends TestCase
+{
+
+    private function validateResult(Result $result, $expectedResult)
+    {
+        $this->assertInstanceOf(Result::class, $result);
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This assertion is unnecessary, due to the typehint; PHP will raise an error for an invalid `$result`, which will cause a failed test.

File: test/BarcodeTest.php [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+        $this->assertSame(
+            $expectedResult,
+            $result->isValid(),
+            'Failed value: ' . var_export($result->getValue(), true)
```
Comment (weierophinney) [inline test/BarcodeTest.php]: Let's make this more descriptive:

```php
sprintf(
    'Expected validation result "%s" was not received; value validated: %s',
    var_export($expectedResult, true),
    var_export($result->getValue(), true)
)
```

File: test/BarcodeTest.php [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+        );
+    }
+
+    public function provideBarcodeConstructor()
+    {
+        return [
+            'null' => [null, Ean13::class],
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This test case will need to be removed once the `AdapterInterface` typehint is added to the `$adapter` constructor argument. As such, the entire data provider can likely be removed, and, instead, the `testBarcodeConstructor()` test case can use a mock instance of `AdapterInterface` to test it.

File: test/BarcodeTest.php [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+        $this->assertInstanceOf($expectedResult, $barcode->getAdapter());
+    }
+
+    public function testNoneExisting()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('not found');
+        $barcode = new Barcode('\Zend\Validate\BarcodeTest\NonExistentClassName');
+        $barcode->getAdapter();
+    }
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This test case can be removed once the updates to the constructor are made.

File: test/BarcodeTest.php [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+
+    public function testSetCustomAdapter()
+    {
+        $barcode = new Barcode($this->createMock(AdapterInterface::class));
+
+        $this->assertInstanceOf(AdapterInterface::class, $barcode->getAdapter());
+    }
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This test will duplicate the `testBarcodeConstructor()` test once the constructor argument is changed, and can thus be removed.

File: test/BarcodeTest.php [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+    {
+        return [
+            'test-set-adapter-upca'  => ['upca', '065100004327', true],
+            'test-set-adapter-ean13' => ['ean13', '0075678164125', true],
+            // @ZF-4352
+            'non-string-validation-upca-int'    => ['upca', 106510000.4327, false],
+            'non-string-validation-upca-array'  => ['upca', ['065100004327'], false],
+            'non-string-validation-ean13-int'   => ['ean13', 106510000.4327, false],
+            'non-string-validation-ean13-array' => ['ean13', ['065100004327'], false],
+        ];
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This provider should be updated to provide actual instances of the adapters.

File: test/_files/MyBarcode1.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator\Barcode;
```
Comment (weierophinney) [inline test/_files/MyBarcode1.php]: Use the `ZendTest\DataValidator\Barcode` namespace, and then import the `AbstractAdapter`.

File: test/_files/MyBarcode2.php [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator\Barcode;
+
+class MyBarcode2 extends AbstractAdapter
```
Comment (weierophinney) [inline test/_files/MyBarcode2.php]: Same comment applies here.

File: test/_files/MyBarcode3.php [hunk 17]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator\Barcode;
```
Comment (weierophinney) [inline test/_files/MyBarcode3.php]: and here, and with all of these test adapters.

File: test/BarcodeTest.php [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+        $input,
+        bool $expectedResult
+    ) {
+        $validator = new Barcode($adapter);
+        $this->validateResult($validator->validate($input), $expectedResult);
+    }
+
+    public function testInvalidChecksumAdapter()
+    {
+        require_once __DIR__ . "/_files/MyBarcode1.php";
```
Comment (weierophinney) [inline test/BarcodeTest.php]: These `require_once` calls can go away once you update the various test adapters to use the `ZendTest` namespace. Doing so will require adding a new `autoload-dev` rule to `composer.json` so they may be autoloaded.

File: test/BarcodeTest.php [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+            'non-string-validation-ean13-int'   => ['ean13', 106510000.4327, false],
+            'non-string-validation-ean13-array' => ['ean13', ['065100004327'], false],
+        ];
+    }
+
+    /**
+     * @dataProvider validationProvider
+     */
+    public function testValidateReturnsExpectedResults(
+        $adapter,
```
Comment (weierophinney) [inline test/BarcodeTest.php]: Typehint this argument.

File: test/BarcodeTest.php [hunk 20]
```diff
@@ -0,0 +1,514 @@
...
+        bool $expectedResult
+    ) {
+        $validator = new Barcode($adapter);
+        $this->validateResult($validator->validate($input), $expectedResult);
+    }
+
+    public function testInvalidChecksumAdapter()
+    {
+        require_once __DIR__ . "/_files/MyBarcode1.php";
+        $barcode = new Barcode('MyBarcode1');
```
Comment (weierophinney) [inline test/BarcodeTest.php]: Here and throughout this file, you'll need to switch from using string "short" names to `new <CLASSNAME>`.

File: test/BarcodeTest.php [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+
+    public function testInvalidAdapter()
+    {
+        $barcode = new Barcode('Ean13');
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('does not implement');
+        require_once __DIR__ . "/_files/MyBarcode5.php";
+        $barcode->setAdapter('MyBarcode5');
+    }
```
Comment (weierophinney) [inline test/BarcodeTest.php]: This method can be removed after the constructor changes are made, and the `setAdapter()` method is removed.

File: test/BarcodeTest.php [hunk 22]
```diff
@@ -0,0 +1,514 @@
...
+
+    public function testCODE25()
+    {
+        $barcode = new Barcode('code25');
+        $this->validateResult($barcode->validate('0123456789101213'), true);
+        $this->validateResult($barcode->validate('123'), true);
+        $this->validateResult($barcode->validate('123a'), false);
+
+        $barcode->useChecksum(true);
```
Comment (weierophinney) [inline test/BarcodeTest.php]: Once the change to `useChecksum()` is made in the `Barcode` class, you'll instead change the flag on the _adapter_ within these tests.

**Comment by harikt** [comment]
Thank you for the detailed review @weierophinney . I will send an update in the coming days.

File: src/Barcode.php [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+
+            if (! class_exists($adapter)) {
+                throw new Exception\InvalidArgumentException('Barcode adapter matching "' . $adapter . '" not found');
+            }
+
+            $adapter = new $adapter();
+
+            if (! is_null($this->useChecksum)) {
+                $adapter->useChecksum($this->useChecksum);
+            }
```
[the remark this answers is not in the document]
Comment (harikt) [reply src/Barcode.php]: Hi @weierophinney , 

If we are passing the concrete adapter, do we really need to pass the `$useChecksum` . There is no place where this variable need to be used. Can't we remove this ?

[the remark this answers is not in the document]
Comment (weierophinney) [reply src/Barcode.php]: If the main barcode validator doesn't and/or shouldn't override what's in the adapter, then we can remove it.

**Comment by harikt** [comment]
Hi @weierophinney ,

Have updated code as requested and added documentation ( copied from old barcode validator , updated with the current ones )

Thank you.

I will add the documentation

File: src/Bitwise.php [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+ * @see       https://github.com/zendframework/zend-datavalidator for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator;
+
+use Traversable;
+
+class Bitwise extends AbstractValidator
```
Comment (harikt) [inline src/Bitwise.php]: Oh shit, I committed this file accidentally 👎  . @weierophinney what do you want to do about this ?

Remove this file from this PR and send a different PR . 

or add unit test / fix for this Bitwise one ?

Comment (weierophinney) [reply src/Bitwise.php]: Remove it.

I've taken the liberty of rebasing your branch off of current master, and I did an interactive rebase to remove the `Bitwise` class file from the commit where it was introduced.

Comment (harikt) [reply src/Bitwise.php]: So do I want to do anything here ? I am a bit confused with the sentence.

> I've taken the liberty of rebasing your branch off of current master, and I did an interactive rebase to remove the Bitwise class file from the commit where it was introduced.

So did you removed the `Bitwise` class or do you want me to remove it ? or when you are doing the rebasing you will do ?

**Comment by harikt** [comment]
It looks like you have removed the Bitwise class. Is there anything else that makes it a blocker for not merging ? I got only one email, but this one still shows requested changes. Not sure what are the changes requested newly. I assume all changes have been already integrated earlier itself. Let me know if I have missed anything.

**Comment by weierophinney** [comment]
@harikt Sorry for the delay getting back to you!

I'd started on some refactoring, but then had other obligations last week that prevented me from finishing, as well as providing you details of what I was doing.

What I've done, in general, is as follows:

- Removed all public setters, in favor of constructor arguments. This ensures that the various barcode adapters, and the `Barcode` validator itself, remain stateless (one of the primary goals of the entire component!).
- I've separated out the "checksum" functionality into an additional interface and trait. Adapters that provide checksum capabilities now implement that interface (optionally via the trait). The `Barcode` validator checks to see if the adapter implements that interface, and if a checksum validation is required, before attempting to validate the checksum. Checksum methods were moved either into the specific adapters, or, if used in more than one location, into traits. This change greatly simplified the `AbstractAdapter`.
- Added `declare(strict_types=1)` throughout all new classes. This helped me identify and correct some problems with length, character, and checksum validations that have been lingering since they were originally written.

I've also updated the documentation to list out any constructor arguments, and to push the usage example to the top of the page.

Thanks a ton for your work on this!

**Comment by harikt** [comment]
> Sorry for the delay getting back to you!

No issues. 

The problem with too much delay is we all will be carried away to other stuffs and forget about what we did and why we did ( Probably a failure at my end to document things ) . I always forget in 2-3 days and find it hard to get back to it. This is not a complaint. Just telling about my personal thing.

And thank you for your good work here. Any plans about release date etc ?
